setup Hitchwiki with ansible

traumschule opened the pull request:
The [ansible](https://ansible.com) playbook (_[hitchwiki.yml](https://github.com/traumschule/hitchwiki/blob/ansible/scripts/ansible/hitchwiki.yml)_) is a major step to make [Hitchwiki v3 production ready](https://github.com/Hitchwiki/hitchwiki/milestone/4). This [branch](https://github.com/traumschule/hitchwiki/tree/ansible) aims for a stable and easy installation process. For future development see  #172.

[![Build Status](https://travis-ci.org/traumschule/hitchwiki.svg?branch=ansible)](https://travis-ci.org/traumschule/hitchwiki)

# Test it!
```
git clone https://github.com/traumschule/hitchwiki -b ansible
./scripts/vagrant/install.sh
./scripts/deploy_remote.sh HOST
```

For details see [INSTALL.md](https://github.com/traumschule/hitchwiki/blob/ansible/INSTALL.md) and [ansible/README.md](https://github.com/traumschule/hitchwiki/blob/ansible/scripts/ansible/README.md).

# Changelog
## _Nov 18_ Install Mediawiki without errors
Run `ansible-playbook hitchwiki.yml` until _Import Mediawiki content_, check debug output and fix the errors (See `mw_import.yml`)
- [X] Run Mediawiki install script (`LocalSettings.php` must be absent, or  [Cannot modify header information - headers already sent](https://www.mediawiki.org/wiki/Manual:Errors_and_symptoms#Warning:_Cannot_modify_header_information_-_headers_already_sent_by_(...)) is risen) [how to find BOMs](https://stackoverflow.com/questions/204765/elegant-way-to-search-for-utf-8-files-with-bom) [backtrace](https://phabricator.wikimedia.org/T157392)
- [X] Setup database for several MW extensions (SemanticMediaWiki, AntiSpoof etc.)
- [X] Pre-populate the AntiSpoof extension's table
- [X] Create MediaWiki users
- [X] Import Semantic templates and other MediaWiki special pages
- [X] the MW update role has several errors, sometimes. automatic error correction would be great.
- [X] add security release: `mediawiki-1.29.2.tar.gz` (November 2017)

## _Nov 13_ secure production setup (fixes #142)
- [X]  [Travis](https://travis-ci.org/) [passes](https://travis-ci.org/traumschule/hitchwiki)
- [X] test next ubuntu and debian with travis
- [X] make ansible branch clean to install via vagrant in one step
- [X] check if necessary variables are set: domain, env, admin_contact, dir.*, mediawiki.protocol, ...
- [X] update [scripts](https://github.com/Hitchwiki/hitchwiki/tree/master/scripts#generic-scripts) _INSTALL.md_
- [X] implement tests to speed up the installer on reruns
- [X] daily dump with backupninja
- [X] #105 cache
- [X] #109 Setup cron jobs
- [X] #138 Override default user password
- [X] README.md: change travis badge to hItchwiki/master
- [X] use [ansible-pull](https://docs.ansible.com/ansible/latest/ansible-pull.html) to deploy

## _Oct 28_ deploy [beta.hitchwiki.org](https://beta.hitchwiki.org) with ansible
- [X] [Monit](https://docs.ansible.com/ansible/latest/monit_module.html): apache, mysql, parsoid, ssh
- [X] #127 https: Certbot / [lets_encrypt module](https://docs.ansible.com/ansible/latest/letsencrypt_module.html)
- [X] Look into unix [ACL](https://docs.ansible.com/ansible/latest/acl_module.html) stuff in `roles/mw/tasks/main.yml`
- [X] Move configs to yaml file (instead of ini file) **https://github.com/Hitchwiki/hitchwiki/pull/164#issuecomment-337671250**
- [X] Test loading  settings with shyaml in `configs/mediawiki.php` and `public/robots.txt.php`
       (alternatives: [spyc](https://github.com/mustangostang/spyc) and [below](#issuecomment-337671250))
- [X] merge roles that won't be distributed to different hosts (eg. web mw dev), outsource groups of task to apps specific files, included conditionally (`when` clause)
- [X] auomate apache configuration
- [X] Fix Apache errors
  - [X] Apache: `/var/www/public/.htaccess: Invalid command 'php_value'` although [mod_php](https://secure.php.net/manual/en/configuration.changes.php) is loaded => comment both lines and add them to /etc/php/7.0/cli/php.ini
  - [X] Apache: `Could not reliably determine the server's fully qualified domain name` although `/etc/apache2/sites-enabled/hitchwiki.conf`defines it => add ServerName outside of <VirtualHost
  - [X] Apache: index.php MWException from line 1205 of /home/hitchwiki/hitchwiki/public/wiki/includes/WebRequest.php: Unable to determine IP. => [fix](https://webmasters.stackexchange.com/questions/19487/mediawiki-problem): add `$_SERVER['REMOTE_ADDR'] = 'YOUR IP ADDRESS';` to `Local Settings.php`
- [X] Fix Mediawiki import errors
- [X] Configure Mediawiki logs path, now: `$logDir = '/vagrant/logs'` (`./configs/mediawiki.php`) 
- [X] test TLS for staging (see INSTALL.md)
- [X] run [post install scripts](https://github.com/Hitchwiki/hitchwiki/blob/master/scripts/install_funcs.sh#L301-L316)
- [X] add script to update LocalSettings.php

Commit:
ansible: first draft

Commit:
scripts/status.sh

Commit:
ansible: basic structure

Commit:
ansible: configure roles, use ansible for provisioning, update README-vagrant.md

simison commented:
Note to myself: add [Travis](https://travis-ci.org/) integration with `ansible-playbook hitchwiki.yml --syntax-check`.

simison commented on INSTALL-vagrant.md:
```diff
@@ -22,29 +22,36 @@ cp configs/vagrant-example.yaml configs/vagrant.yaml
+When you're ready to begin working again, just run `vagrant provision`.
+
+#### Ansible
+As soon as ssh is ready, [Ansible](https://docs.ansible.com/ansible/latest/intro.html) takes over to configure your system according to the `hitchwiki.yml` [Playbook](https://docs.ansible.com/ansible/latest/playbooks_intro.html):
```
_"As soon as ssh is ready"_ => _"As soon as Vagrant machine is ready"_ probably more clear for anyone new to Vagrant.

traumschule replied:
done

simison commented on group_vars/vars:
```diff
@@ -0,0 +1,22 @@
+VAGRANT_WIKIDIR: "{{ VAGRANT_ROOTDIR }}/public/wiki"
+
+# Hitchwiki
+hw_owners: "{{ HW__general__webserver_user }}:{{ HW__general__webserver_group }}"
```
Note that `HW__general__webserver_user` [were set with bash](https://github.com/Hitchwiki/hitchwiki/blob/master/scripts/_settings.sh) from `settings.ini`, but here they might need to be set just as strings? They could be taken out form settings.ini as this doesn't really need to be configurable. `www-data:www-data` is the default Apache2 user:group.

traumschule replied:
TODO as soon as i figure out, how to include them from a script, they will depend on _settings.sh, or if we decide to deprecate it, from settings.ini.

simison replied:
I think there's a bunch of scripts now relying on `_settings.ini`, although those can probably all be migrated to Ansible.

`settings.sh` setting up env vars for Ansible could be [one trick](https://docs.ansible.com/ansible/latest/intro_configuration.html#getting-the-latest-configuration), although in general I always feel like they are a bit of a blackbox so I try to avoid them.

simison commented on roles/common/handlers/main.yml:
```diff
@@ -0,0 +1,28 @@
+     # TODO use scripts/status.sh
+     command: |
+       echo "Hitchwiki is now installed!
+       Apache is up. Open http://{{ HW__geieral__domain }}/ in your browser.
```
Note typo in `HW__geieral__domain`.

traumschule replied:
fixed

simison commented on roles/common/tasks/main.yml:
```diff
@@ -0,0 +1,27 @@
+   - name: install helper tools
+     apt: package={{ item }} state=latest
+     with_items:
+       - python
```
Python was installed in pre-tasks already, right?

traumschule replied:
thanks, removed it

simison commented on roles/db/tasks/main.yml:
```diff
@@ -0,0 +1,46 @@
+    content: |
+      [client]
+      user={{ HW__db__username }}
+      password={{ HW__db__password }}
```
This is for dev env only, right?

traumschule replied:
with the next commit the whole section is moved to roles/dev/tasks/main.yml

simison commented on roles/mw/tasks/main.yml:
```diff
@@ -0,0 +1,117 @@
+  # Config file is stored elsewhere, require it from MW's LocalSettings.php
+  copy:
+    src: "{{ SCRIPTDIR }}/configs/mediawiki_LocalSettings.php"
+    dest: "{{ WIKIDIR }}/LocalSettings.php"
```
Let's keep this as a symlink so that we can pull modifications to configuration file `/configs/mediawiki_LocalSettings.php` without needing to go re-copy it to `public/wiki/LocalSettings.php`.

simison replied:
Oh wait, that file just import/requires it? So ignore what I just said. :-)

traumschule replied:
done

simison commented on roles/web/tasks/main.yml:
```diff
@@ -0,0 +1,91 @@
+    sed -i "s/error_reporting = .*/error_reporting = E_ALL/" /etc/php/7.0/apache2/php.ini
+    sed -i "s/display_errors = .*/display_errors = On/" /etc/php/7.0/apache2/php.ini
+  notify:
+    - restart apache
```
This should be off by default (at least in prod) and we have [a flat at config](https://github.com/Hitchwiki/hitchwiki/blob/master/configs/mediawiki.php#L59-L63) we use to control this.

traumschule replied:
moved to dev as well now

simison replied:
It should still be off by default, so that setting `$hwDebug=false` will hide them.

Commit:
ansible: minor fixes

Commit:
ansible: add _path_resolve.sh _settings.sh and settings.ini to group_vars/all

simison commented:
@guaka or @omelnyk do you know how to read variables into Ansible from external config files (yaml/ini/json/whatnot)?

guaka commented:
Just from .yml files

On Wed 18. Oct 2017 at 18:57, Mikael Korpela <notifications@github.com>
wrote:

> @guaka <https://github.com/guaka> or @omelnyk <https://github.com/omelnyk>
> do you know how to read variables into Ansible from external config files
> (yaml/ini/json/whatnot)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Hitchwiki/hitchwiki/pull/164#issuecomment-337657160>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADVyENJw6zqUp8M6e-9szaPX7z313Qhks5sti3ygaJpZM4P8Pb6>
> .
>
-- 
sent from my phone

simison commented:
Aight!

@traumschule so I guess [this line in Mediawiki config](https://github.com/Hitchwiki/hitchwiki/blob/master/configs/mediawiki.php#L26) needs to use [PHP YAML](https://secure.php.net/manual/en/book.yaml.php) then.

As for bash files, might be we won't have those left after Ansible but if they need access to settings, there's this: https://github.com/0k/shyaml

simison commented:
Two other yaml readers for php that we can install via Composer:
https://symfony.com/doc/current/components/yaml.html
https://github.com/mustangostang/spyc/

Commit:
ansible: improve roles

Commit:
ansible: improve roles

Commit:
Replace settings.ini with settings.yml

group_vars/all includes
 - configs/settings.yml (link/copy from settings-example.yml)
 - configs/vagrant.yaml

This superseeds and removes the following files:
 - configs/settings-example.ini
 - scripts/_path_resolve.sh
 - scripts/_settings.sh

The following files are replaced by ansible:
 - scripts/server_install.sh
 - scripts/install_funcs.sh
 - scripts/install_parsoid.sh
 - scripts/create_users.sh
 - scripts/install_local.sh
 - scripts/vendor/bash_ini_parser

Commit:
New settings.yml replaces settings.ini

    First copy configs/settings-example.yml to settings.yml

    This commit superseeds the following files:
     - configs/settings-example.ini
     - configs/vagrant-example.yaml
     - configs/vagrant.yaml
     - scripts/server_install.sh

   install_local.sh still depends on:
     - scripts/_path_resolve.sh
     - scripts/_settings.sh
     - scripts/vendor/bash_ini_parse
     - scripts/install_funcs.sh
       - scripts/install_parsoid.sh
       - scripts/create_users.sh

    composer.json references
     - scripts/vagrant/install.sh
     - scripts/install_local.sh
     - scripts/clean.sh

    Links
     - group_vars/all -> group_vars/hitchwiki
     - group_vars/hitchwiki -> configs/settigs.yml
     - configs/vagrant-example.yaml -> configs/settings-example.yml
     - configs/settings-example.ini -> configs/settings-example.yml

Commit:
Add scripts/status.sh

simison commented:
> Use lets_encrypt module (and replace scripts/cert_selfsigned.sh)

cert_selfsigned.sh is still needed for dev environment, because lets encrypt works only for domains which are publicly accessible.

simison commented:
>Update migrations: scripts/bot/*

I think it's okay to let these read `settings.ini`, we'll just know to do that once we need to run the migration bot. Once migration is done, this whole directory is going to get wiped out anyway.

Commit:
Ansible: Re-enable NodeSource, load settings via yml robots.txt.php (untested)

traumschule commented on INSTALL-vagrant.md:
```diff
@@ -14,37 +14,71 @@ git clone https://github.com/Hitchwiki/hitchwiki.git && cd hitchwiki
 
 After setup your virtual machine is running. Suspend the virtual machine by typing `vagrant suspend`.
-When you're ready to begin working again, just run `vagrant up`.
+When you're ready to begin working again, just run `vagrant provision`.
```
ah right, after `suspend`, `up` is necessary.
`scripts/vagrant/insall.sh` does `vagrant up` and `vagrant provision` in one - if it doesn't fail on first run (if we ever get there!)

Commit:
ansible: Add deploy_remote.yml playbook + some improvements

Commit:
ansible: Add status.yml and update roles

Commit:
ansible: add monit

Commit:
ansible: add TLS support

Commit:
Replace outdated INSTALL.md with updated INSTALL-vagrant.md

simison commented on configs/mediawiki.php:
```diff
@@ -23,6 +23,8 @@
 # Load Hitchwiki Config
+# TODO https://secure.php.net/manual/en/book.yaml.php
+#$hwConfig = eval 'cat ../configs/settings.ini | shyaml get-value mapping';
 $hwConfig = parse_ini_file("settings.ini", true);
```
So something like this here should work:
```php
# Load Hitchwiki Config
# Paths relative to `configs/mediawiki.php`
require_once '../public/mustangostang/spyc/spyc.php';
$hwConfig = spyc_load_file('settings.yaml');
```

...just need to check in what structure this loads data from the file, so that we can modify references like `$hwConfig['spam']['spamprotection']` accordingly. Just throwing `die(print_r($hw_config,true));` after these lines shows it.

traumschule replied:
what about installing it into `scripts/vendor/spyc`, or `scripts/vendor/mustangostang/spyc` (easier to .gitignore)?

simison commented on configs/settings-example.yml:
```diff
@@ -0,0 +1,131 @@
+# Variables for ansible provisioning
+# See https://docs.ansible.com/ansible/latest/playbooks_variables.html
+# https://docs.ansible.com/ansible/latest/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable
+env: "dev"
```
Let's ditch this, it's used only twice. I've learned it's a bad practise to have env flags like this, and we should have feature flags instead (like we do, e.g. `debug:true` or `cache:true`).

https://github.com/Hitchwiki/hitchwiki/blob/master/configs/mediawiki.php#L57

That section can be like this instead:
```php

// Enable error reporting
if ($hwDebug) {
  error_reporting( -1 );
  ini_set( 'display_errors', 1 );
  
  // Show the debug toolbar if `hw_debug` is set on the request,
  // either as a parameter or a cookie.
  // https://www.mediawiki.org/wiki/Debugging_toolbar
  if ( !empty( $_REQUEST['hw_debug'] ) ) {
    $wgDebugToolbar = true;
  }
}

// Expose debug info for PHP & SQL errors.
$wgShowExceptionDetails = $hwDebug;
$wgDevelopmentWarnings = $hwDebug;
$wgDebugDumpSql = $hwDebug;
$wgShowDBErrorBacktrace = $hwDebug;
$wgShowSQLErrors = $hwDebug;
$wgResourceLoaderDebug = $hwDebug;

// Profiling
$wgDebugProfiling = false;

// Log into file
$logDir = '/vagrant/logs'; //TODO, 
$wgDebugLogFile = "{$logDir}/mediawiki-debug.log";
foreach ( array( 'exception', 'runJobs', 'JobQueueRedis' ) as $logGroup ) {
  $wgDebugLogGroups[$logGroup] = "{$logDir}/mediawiki-{$logGroup}.log";
}
```

https://github.com/Hitchwiki/hitchwiki/blob/ba64070be06999c43410f5d56fc2fad1bccd8566/public/robots.txt.php#L12

I'll write more about robots.txt later.

simison replied:
(Can be dealt also later on, though!)

traumschule replied:
php is not my greatest fun, i leave this up to you :) #166

simison commented on configs/settings-example.yml:
```diff
@@ -0,0 +1,131 @@
+  helper: [ unattended-upgrades, composer, tmux, nodejs, npm, python-pip, vim, curl, git, unzip, zip, imagemagick, build-essential, python-software-properties, fail2ban ]
+  db: [ mariadb-server, software-properties-common, python-mysqldb ]
+  web: [ apache2, phpmyadmin, php7.0, libapache2-mod-php7.0, php7.0-mysql, php7.0-curl, php7.0-gd, php7.0-intl, php7.0-imap, php7.0-mcrypt, php7.0-pspell, php7.0-recode, php7.0-sqlite3, php7.0-tidy, php7.0-xmlrpc, php7.0-xsl, php7.0-mbstring, php7.0-opcache, php-memcache, php-pear, php-imagick, php-apcu, php-gettext ]
+  mw: [ apt-transport-https, parsoid ]
```
Shouldn't these be directly at the ansible book, or updating versions/setup would be quite much more harder?

traumschule replied:
right, let ansible abstract away system stuff.

simison commented on configs/settings-example.yml:
```diff
@@ -0,0 +1,131 @@
+webserver_user: www-data
+webserver_group: www-data
+
+hw:
```
Let's have this as "mediawiki" instead of "hw" for clarity.

traumschule replied:
got me again being lazy :)

what about this structure:
```
dir
  root
  ...

ansible
  environment
  user
  status
    db
    mw
    ...
  version
    apache
    npm
    ...
  
vagrant
  private_network_ip
  hostname
  hostmanager_enabled

db
   ...

hitchwiki
  protocol: http|https # => used by ansible to (de)active transport security
  smtp
    ...
  spam
    ...
  mapbox
    ...
  geonames_usernames: hitchwiki{,2,3}
```

simison commented on public/robots.txt.php:
```diff
@@ -6,7 +6,8 @@
+$hwConfig = cat ../configs/settings.ini | shyaml get-value mapping
 
 // Production environment
 if(isset($hwConfig['general']['env']) && $hwConfig['general']['env'] == 'prod'):
```
Instead of `env` config, we can add extra feature config variable to:
```php
if($hwConfig['general']['visibleForSearchEngines'])):
```

simison replied:
Again kinda unrelated so I can do this in a separate PR as well!

simison commented on roles/dev/tasks/main.yml:
```diff
@@ -0,0 +1,107 @@
+    content: |
+      [client]
+      user={{db.username}}
+      password={{db.password}}
```
What process is using this file?

traumschule replied:
if it's there, all mysql commands, but I'll drop it in favor of vault (encrypted variable storage). question then is how to share the password between developers in the future.

simison replied:
We don't have too many situations when we'd have to enter SQL credentials (only when provisioning the production, really). For the dev env these can be just something what we can push to git, like root/root.

simison replied:
I.e. keeping it simple might be better. ;-)

simison commented on roles/dev/tasks/main.yml:
```diff
@@ -0,0 +1,107 @@
+    regexp: '^error_reporting = .*'
+    line: 'error_reporting = E_ALL'
+  when: debug
+  notify: restart apache
```
Let's have PHP error reporting _off_ by default (it should already be?) so that we can turn it _on_ then on _runtime only_ (at mediawiki.php). Kinda just more secure practise in that way. :-)

Commit:
Load Mediawiki config from settings.yml

Commit:
ansible: Turn off PHP error reporting by default

traumschule commented:
all mysql commands, but this is no longer needed as all calls from
ansible will use vault in the future

Commit:
ansible: don't save credentials to disk during development

simison committed:
Merge branch 'master' into ansible

Commit:
Add cron job for daily localisation updates (Fixes #109)

Commit:
Merge branch upstream/master into ansible

Commit:
ansible: Make wiki users configurable (Fixes #138)

Commit:
ansible: fix apache errors

simison commented on roles/dev/tasks/setup_ssl.yml:
```diff
@@ -0,0 +1,142 @@
+  shell: |
+    cd {{ROOTDIR}}
+    if [[ $* == *--ssl* ]]; then
+      bash {{SCRIPTDIR}}/cert_selfsigned.sh
```
BTW I just stumbled upon this: https://www.jeffgeerling.com/blog/2017/self-signed-certificates-ansible-local-testing-nginx — might be helpful.

simison committed:
Fix syntax error at settings.ini

simison commented on hosts:
```diff
@@ -0,0 +1,4 @@
+[hitchwiki]
+192.168.33.10 ansible_user=ubuntu ansible_ssh_private_key_file=/home/hitchwiki/src/hitchwiki/.vagrant/machines/hitchwiki/virtualbox/private_key
```
This probably shouldn't be in git?

simison committed:
Update / clarify documentation

simison commented on scripts/vagrant/install.sh:
```diff
@@ -4,20 +4,33 @@ cd "$(dirname $0)/.."
-echo ""
+echo -e $sep
+
+echo "Checking system ..."
```
We'll have so many different systems where this file might be run, that it's better to just have these steps listed at INSTALL.md's "Prerequisites" section. I just added Ansible there.

E.g. I'm running MacOS and this just throws a bunch of errors for me. :-)

traumschule replied:
just pushed a commit fixing the bash issue

Commit:
vagrant/install.sh: Don't install software directly.

Commit:
ansible: merge roles into 'hitchwiki' role

Commit:
ansible: clean apt dependencies from settings

Commit:
ansible: clean apt dependencies from settings files + other fixes

Commit:
ansible: give handlers more intuitive names

Commit:
ansible: fix notify

Commit:
ansible: restructure variables, debug output for mediawiki import

simison committed:
Add `visible_to_search_engines` setting to control robots.txt.php

simison committed:
Move comments to newlines at settings file

simison committed:
Move prefix setting to mediawiki section at settings

simison committed:
Mediawiki config: Remove dependency to env setting and load configuration from yaml file

simison committed:
Fix syntax error at roles/hitchwiki/tasks/main.yml

simison committed:
Clean out Vagrant config

simison committed:
Remove unnessessary pip and bundler dependencies from scripts/vagrant/install.sh

simison committed:
Sort out clean.sh scripts

simison committed:
Clarify prerequisites at install.md

simison committed:
Add missing . to clean command at install.md

Commit:
separate settings from application specific variables

simison commented on configs/settings-example.yml:
```diff
@@ -0,0 +1,55 @@
+
+vagrant:
+  private_network_ip: "192.168.33.10" # Sets IP address for the Virtual Machine
+  hostname: "{{ mediawiki.domain }}" # Hostname to use for the development server
```
Since these settings are read by Vagrant at this point, this won't work (`{{ mediawiki.domain }}` is an Ansible feature?) Normally yaml files can't have logic like this.

Now Vagrant throws:
```
There are errors in the configuration of this machine. Please fix
the following errors and try again:

vm:
* The hostname set for the VM should only contain letters, numbers,
hyphens or dots. It cannot start with a hyphen or dot.
```

Let's move the hostname to top level since it's a shared setting used by Vagrant, Mediawiki and Apache.

traumschule replied:
I assume  `domain` is identical to `hostname`, sw we can get rid of mediawiki.domain.

simison commented on configs/settings-example.yml:
```diff
@@ -0,0 +1,55 @@
+    Hitchwiki: "{{ db.phpmyadmin_password }}"
+    Hitchbot: "{{ db.phpmyadmin_password }}"
+    Hitchhiker: "{{ db.phpmyadmin_password }}"
+  smtp: # MailDev is running at port 1025 at dev server https://github.com/djfarrelly/MailDev
```
SMTP settings could also be at top level since they're gonna be the same for anything else than just MediaWiki.

traumschule replied:
done

simison commented on configs/settings-example.yml:
```diff
@@ -0,0 +1,55 @@
+# Variables for ansible provisioning
+#  https://docs.ansible.com/ansible/latest/playbooks_variables.html
+
+env: "dev" # dev | production - to enable TLS change mediawiki.protocal to 'https' below
```
Let's rather have a separate "ssl: True` so that we can drop the whole `env` thing.

traumschule replied:
ssl is deprecated.

simison replied:
Huh?

traumschule replied:
"SSL 2.0 was prohibited in 2011 by RFC 6176, and SSL 3.0 was deprecated in June 2015 by RFC 7568" https://en.wikipedia.org/wiki/Transport_Layer_Security

simison commented on .gitignore:
```diff
@@ -8,6 +8,11 @@ node_modules/
 .vagrant
 
+# Ansible
+./group_vars
```
This can probably be removed from here?

traumschule replied:
not sure yet

simison commented on .gitignore:
```diff
@@ -8,6 +8,11 @@ node_modules/
 
+# Ansible
+./group_vars
+hosts
```
Seems like `hosts` file needs to exist or Ansible won't even start.

We might very well keep some defaults like:
```
[local]
127.0.0.1

[production]
hitchwiki.org

[beta]
beta.hitchwiki.org
```
...but I suppose it needs to allow adding custom targets during development? — same trick as with settings.yml?

traumschule replied:
hosts is defined in ansible.cfg, to get rid of hosts, delete both.

To make your sample work i need to swap:
- `group_vars/all` - is available to all and currently has
- `group_vars/hitchwiki` links to configs/settings.yml and is only available to hosts in the group [hitchwiki]

My hosts has
```
[hitchwiki]
vagrant ansible_ssh_host=127.0.0.1 ansible_user=ubuntu ansible_ssh_port=2222 ansible_ssh_private_key_file=/home/hitchwiki/src/hitchwiki/.vagrant/machines/hitchwiki/virtualbox/private_key
hw-dev ansible_ssh_host=... ansible_user=hitchwiki
```

To set variables per host use `host_vars/name`. My `host_vars/vagrant` has "apt_proxy: pi".

simison commented on roles/hitchwiki/tasks/system.yml:
```diff
@@ -0,0 +1,64 @@
+  pip: name=shyaml state=present chdir={{ dir.script }}/vendor extra_args=--user
+
+- name: Add daily cron job to update system
+  cron: name=apt special_time=daily job="apt-get update && apt-get dist-upgrade -yq && apt-get autoremove -yq"
```
We can just install this?
```
sudo apt-get install unattended-upgrades
```

simison commented on roles/hitchwiki/tasks/system.yml:
```diff
@@ -0,0 +1,64 @@
+      # See https://blog.packagecloud.io/eng/2015/05/05/using-apt-cacher-ng-with-ssl-tls/
+  when: apt_proxy is defined
+
+- name: Remove apt source of NodeSoure 8.x # to avoid dependency problems
```
What problems did this cause? Seems like this line is later being put back at Parsoid installation?

traumschule replied:
yes, npm can't be installed when these sources are present.

simison replied:
Interesting. Would it work with this:
```
curl -sL https://deb.nodesource.com/setup_8.x | sudo -E bash -
sudo apt-get install -y nodejs
```
[via](https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux-distributions)

i.e. install nodejs+npm, instead of just npm.

It would be better to have lines referring to node -version all in one place, so that it would be easier to update later on.

Also Maildev and Bower rely on Node.js so it makes sense to install it already here, instead of only at Parsoid step.

traumschule replied:
We are not installing with the bash script, don't really see a different to the repo in your code:
```
- name: Add apt source for NodeSoure 8.x
  become: yes
  apt_repository: filename=nodesource.list state=present repo="deb https://deb.nodesource.com/node_8.x xenial main"

- name: Install nodejs
  become: yes
  apt: name=nodejs state=present update_cache=yes

- name: Link nodejs to node
  become: yes
  file: state=link src=/usr/bin/nodejs path=/usr/local/bin/node

$ sudo apt-get install nodejs npm
Reading package lists... Done
Building dependency tree       
Reading state information... Done
npm is already the newest version (3.5.2-0ubuntu4).
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 nodejs : Conflicts: npm
E: Unable to correct problems, you have held broken packages.
```

simison commented on roles/hitchwiki/tasks/system.yml:
```diff
@@ -0,0 +1,64 @@
+      export LC_ALL="en_US.UTF-8"
+      export LANGUAGE="en_US.UTF-8"
+
+- name: Use apt-cacher
```
What's this used for?

traumschule replied:
apt proxy that caches apt files, useful for slow internet when you repeatedly destroy your box

Commit:
ansible: use facts.d; use monit; update settings-example.yml

Commit:
ansible: minor fixes to mw_import.yml

Commit:
restore scripts/init_maildev.sh

Commit:
ansible: Configure monit, add links to ./logs

Commit:
mediawiki.php: fix spyc path

Commit:
fix mediawiki arrays

Commit:
ansible.cfg

Commit:
configs/monitrc - log

Commit:
monit: set poll time

Commit:
restore scripts/_path_resolve.sh

Commit:
update configs/mediawiki-extensions.php

Commit:
scripts/logs.sh - show all logs with less

Commit:
composer: make MW branch configurable

Commit:
fix MW import, mediawiki.php templates to LocalSettings.php

Commit:
drop settings-example.ini and bash_ini_parser

Commit:
scripts/update.sh: add playbook to update LocalSettings.php

Commit:
let ansible manage TLS

Commit:
ansible: sercure TLS setup

Commit:
ansible: finish first release

Commit:
ansible 0.1

Commit:
ansible: already install nodejs in system role

simison commented on roles/remote/vars/main.yml:
```diff
@@ -0,0 +1,2 @@
+user: hitchwiki
+hostname: hitchwiki.test
```
what does this hostname refer to? Note that it's on config file which is pushed to repo, so it's kind of not configurable.

traumschule replied:
it was no longer used, dropped the file

simison commented on configs/monitrc:
```diff
@@ -0,0 +1,120 @@
+# group services
+# start program = "/etc/init.d/maildev start" with timeout 60 seconds
+# restart program  = "/etc/init.d/maildev restart"
+# stop program  = "/etc/init.d/maildev stop"
```
Since we're not going to run maildev on prod, this could be removed?

Commit:
ansible: don't create bot users for production

Commit:
ansible: add some helpers

Commit:
ansible: update group_vars

Commit:
ansible: add user option, drop 'roles/remote/vars/main.yml'

Commit:
ansible: monit should not monitor maildev, disable monit module - ansible:#30614

Commit:
ansible: configure backupninja, add bot users only for development

Commit:
ansible: secure dbserver, tests, stop_all.sh, see INSTALL.md

Commit:
ansible: scripts/clean.sh must not delete group_vars/; fix settings-example.yml

Commit:
ansible: make deployment more smooth

Commit:
ansible: more deployment fixes

Commit:
move ansible stuff to 'scripts/ansible'

Commit:
ansible: update links; add REAMDE for *_vars/

Commit:
ansible: structure roles deploy update status

Commit:
deploy script for remote and vagrant

Commit:
ansible: update INSTALL.md, improve deployment

Commit:
ansible: faster status check

Commit:
ansible: better role workflow, fix db security

Commit:
ansible: syntax fixes

Commit:
ansible: TLS checks

Commit:
ansible: vagrant settings, require user to ssh in for installation

Commit:
ansible: fix parsoid, monit, status.sh

Commit:
ansible: production status check

Commit:
ansible: hitchwiki.yml

Commit:
ansible: run handler 'update system' with allow_unauthenticated=yes

Commit:
drop group_vars, shorten status role

Commit:
ansible: add discourse installation

Commit:
ansible: add swapfile support, disable some options in vagrant

Commit:
ansible: unignore hosts

Commit:
ansible: add default hosts for vagrant

Commit:
ansible: standard vagrant+ansible config

Commit:
ansible: use port 2222 for vagrant+ansible

Commit:
ansible: set standard port to 2222

Commit:
ansible: automatically call HW installer after provisioning

Commit:
ansible: drop deploy_remote.yml and fix startup errors

Commit:
ansible: improve output of status_all.sh

Commit:
ansible: add travis tests for hitchwiki role

Commit:
ansible: add option to not delete MW db or folder + more fixes

Commit:
ansible: move .travis.yml to root

Commit:
ansible: cleaner deployment

Commit:
ansible: fix .travis.yml

Commit:
ansible: fix roles_path in .travis.yml

Commit:
ansible: set 'roles_path = ../'  in .travis.yml

Commit:
ansible: roles_path in .travis.yml again

Commit:
ansible: 'roles_path = scripts/ansible/' in .travis.yml

Commit:
ansible: 'roles_path = scripts/ansible/roles/hitchwiki/' in .travis.yml

Commit:
ansible: update INSTALL.md

Commit:
ansible: load settings.yml early

Commit:
ansible: fix mode of wiki folders

Commit:
ansible: add scripts link

Commit:
ansible: improve status checks

Commit:
ansible: enable monit by default, apache TLS reqrite

Commit:
ansible: minor change for reruns

Commit:
ansible: remove TLS from production.yml as it's called from hitchwiki/tasks/main.yml

Commit:
ansible: add scripts/clean_server.sh

Commit:
ansible: fix MW rewrite error

Commit:
ansible: fix .travis.yml

Commit:
ansible: add certbot --agree-tos

Commit:
ansible: add role deploy to tests

Commit:
ansible: ignore_errors=yes for 'update system' handler

Commit:
ansible: minor fix to status.sh

Commit:
ansible: provisory discourse installation with rails server

Commit:
ansible: provisory discourse installation (please update settings.yml)

Commit:
ansible: travis a bit

Commit:
ansible: skip authorized_keys for travis

Commit:
ansible: skip authorized_keys with message with no present public key

Commit:
ansible: skip authorized_keys when none are found

Commit:
ansible: explicitly add universe for travis

Commit:
ansible: require xenial for travis

Commit:
ansible: install composer via php for travis

Commit:
ansible: re-enable universe for travis (php-apcu)

Commit:
ansible: fix template error

Commit:
ansible: ignore missing php-apcu

Commit:
ansible: select mariadb source by OS/codename

Commit:
ansible: allow_unauthenticated=yes for mariadb

Commit:
apache: fix minor syntax error in apache.yml

Commit:
ansible: add php-backports for travis

Commit:
ansible: ignore MW install error on rerun

Commit:
ansible: travis don't ask

Commit:
travis: ignore apt errors

Commit:
travis: make apache core mandatory

Commit:
travis: make PHP7 mandatory

Commit:
travis: ignore php errors

Commit:
travis: try xenial container

Commit:
travis: install apache+php7 in before_install

Commit:
minor fixes to .travis.yml, setup_hichwiki.sh, stop_all.sh

Commit:
travis: switch to latest ubuntu LTS docker container

Commit:
travis: correct container name

Commit:
travis: install ansible in script

Commit:
travis: require software-properties-common

Commit:
travis: run apt-get update first

Commit:
travis: require python-pip

Commit:
ApiSandbox was merged into mediawiki/core https://gerrit.wikimedia.org/r/#/c/209570/

Commit:
travis: add link to roles in tests

Commit:
travis: require openssh-client

Commit:
ansible: install sudo during deployment

Commit:
travis: ignore 'Device or resource busy' when changing hostname

Commit:
ansible: use import_tasks for handlers to work

Commit:
ansible: use include_tasks with dynamic variables

Commit:
ansible: be tolerant about mariadb's service name

Commit:
ansible: install at in deploy

Commit:
travis: fix path

Commit:
ansible: check for synced dir with stat module

Commit:
ansible: fix indention error

Commit:
ansible: add proper maildev initscript, minor tweaks to TLS and parsoid"

Commit:
ansible: improve tls_selfsigned.yml

Commit:
ansible: check if necessary variables are set

Commit:
ansible: skip TLS if not requested, let docker stop after tests passed

Commit:
ansible: update _scripts/README.md_, background maildev service or playbook won't finish

Commit:
ansible: update maildev handler

Commit:
ansible: start services after setup; fix stop_all.sh

Commit:
ansible: flush_handlers more frequently

Commit:
ansible: restart services without handlers, test urls from outside docker

Commit:
ansible: run status before update

Commit:
ansible: create settings.yml if necessary

Commit:
travis: fix timeout error

Commit:
travis: be less verbose

Commit:
travis: create /etc/ansible

Commit:
travis: don't check maildev

Commit:
travis: show pip output

Commit:
travis: quote command

Commit:
travis: syntax fix

Commit:
ansible: set repo to upstream/master (travis badge in README.md)

settings-example.yml now has repository with url and branch,
respected by the deploy role (used by deploy_remote.sh).

Commit:
fix parsoid db prefix in mediawiki-extensions.php

Commit:
ansible: minor changes to hitchwiki role

traumschule commented:
I think this is ready for Hitchwiki master.
![Yay!](https://cloud.githubusercontent.com/assets/481677/11768886/18d68472-a1a0-11e5-90b1-51db36c1ab0e.gif)

Commit:
travis: test multiple OS

Commit:
ansible: add option keyserver=hkp://keyserver.ubuntu.com:80

Commit:
ansible: correct path of monitrc

Commit:
test_hitchwiki_playbook.sh -> test_syntax.sh

Commit:
ansible: easier provisioning

Commit:
ansble: more robust node installer

Commit:
add option keyserver (default=hkp://keyserver.ubuntu.com:80)

Commit:
let phpmyadmin change apache.conf

Commit:
merge branch ansible

Commit:
ansible: setup_hitchwiki.sh now checks that deploy added localhost to inventory

Commit:
quote version strings in status.sh

Commit:
add _scripts/ansible/README.md_

Commit:
ansible: improved status output

Commit:
ansible: ansible/README.md + vagrant provisions with ansible_local

Commit:
ansible: fix status path

Commit:
ansible: make handler 'load status' more robust

Commit:
ansible: become=true to configure maildev

Commit:
travis: use default notification message

Commit:
ansible: set default _dir.state=logs/state.yml_

Commit:
travis: test debian and ubuntu:rolling again

Commit:
ansible: error_on_missing_handler = True

Commit:
test_syntax.sh -> check.sh (ansible+travis syntax checker)

Commit:
travis: refactor

Commit:
make ansible branch stable (disable testing)

Commit:
ansible: add http+https check

Commit:
testing: os_version = ansible_distribution_release || stable

Commit:
testing: Install PHP7.1 for artful and bionic

Commit:
testing: add .circleci

Commit:
ansible: refactor deploy, fix state path

Commit:
Update MW to 1.29.2 (supported until July 2018)

https://www.mediawiki.org/wiki/Release_notes/1.29

Commit:
disable Circle CI for testing and idempotence

Commit:
Drop dependency of ExternalData extension (fixes #114)

Commit:
always drop DB before MW installation

Commit:
add code to load settings with symfony (commented)

Commit:
Drop deploy role + add MW debug messages during install

Commit:
disable maildev for now

Commit:
fix deploy_remote.sh

Commit:
test 127.0.0.1 insted of localhost

Commit:
deploy_remote.sh: Copy local configs/settings.yml and dumps/letsencrypt.tar.xz if present

Commit:
Grant whole webroot to {{user}}.{{apache.group}}, minor permission fix

guaka commented:
This week I put together some ansible code to deploy hitchwiki-private to a new server, it's almost done.

guaka commented:
it's unfortunate but I don't think this will be (re)used - but please step up if you do want to (re)use some of this work
